agent/prototyper.py

```diff
@@ -0,0 +1,228 @@
+"""An LLM agent to generate a simple fuzz target prototype that can build.
+Use it as a usual module locally, or as script in cloud builds.
+"""
+import logging
+import os
+import re
+import subprocess as sp
+import threading
+from typing import Optional
+
+from agent.base_agent import BaseAgent
+from llm_toolkit.prompt_builder import DefaultTemplateBuilder
+from llm_toolkit.prompts import Prompt
+from result_classes import BuildResult, Result
+from tool.container_tool import ProjectContainerTool
+
+# TODO(dongge): Replace this with google-cloud-log in a module.
+# Custom logging filter to add thread-local information to log records
+thread_local = threading.local()
+
+#TODO(dongge): Replace process ID with trial ID.
+logging.basicConfig(level=logging.DEBUG,
+                    format=('%(asctime)s [PID: %(process)d] %(levelname)s '
+                            '[%(module)s.%(funcName)s]: %(message)s'))
+
+logger = logging.getLogger(__name__)
+logger.setLevel(logging.DEBUG)
+
+
+def write_to_file(file_path: str, file_content: str):
+  with open(file_path, 'w') as file:
+    file.writelines(file_content)
+
+
+MAX_ROUND = 100
+
+
+class Prototyper(BaseAgent):
+  """The Agent to generate a simple but valid fuzz target from scratch."""
+
+  def _initial_prompt(self, results: list[Result]) -> Prompt:
+    """Constructs initial prompt of the agent."""
+    benchmark = results[-1].benchmark
+
+    default_prompt_builder = DefaultTemplateBuilder(model=self.llm,
+                                                    benchmark=benchmark)
+    prompt = default_prompt_builder.build([])
+    work_dirs = results[-1].work_dirs
+    prompt.save(work_dirs.prompt)
+    return prompt
+
+  def _parse_tag(self, response: str, tag: str) -> str:
```

Comment:
should some of these go into `BaseAgent`? I imagine a lot of these things related to formatting, parsing/executing commands would be required for all agents?

Reply:
Yes, good point, thanks!
I will go through them and move suitable ones to the base class.

Reply:
Done below